### PR TITLE
[Transforms] Adds profiler events for cpu profile analysis of the compiler

### PR DESCRIPTION
**Pull request by rbuckton**
This change adds several profiler events that are picked up by the ts profiler for assist with analyzing cpu profiles of the compiler.

### CODE REVIEW

**Comment by weswigham** [comment]
Is there a particular reason you've decided to make `extendedDiagnostics` require `diagnostics`? I'm pretty sure it's more like `extendedDiagnostics` should imply `diagnostics`, since it's useless without the latter.

File: src/compiler/core.ts [hunk 1]
```diff
@@ -487,6 +487,54 @@ namespace ts {
...
+
+    /**
+     * This is a v8 optimization also used by dart. It uses a heuristic that forces v8 to allocate
+     * a hidden class for an object if it does not already have one.
+     *
+     * Fast maps are better for smaller maps or frequent read operations, but are worse for frequent
+     * insertions of new properties as that causes numerous hidden class transitions and increases
+     * polymorphism.
+     */
+    export function toFastMap<T>(map: Map<T>) {
```
Comment (weswigham) [inline src/compiler/core.ts]: This isn't called anywhere in this PR. Intentional?

File: src/compiler/core.ts [hunk 2]
```diff
@@ -487,6 +487,54 @@ namespace ts {
 
     const hasOwnProperty = Object.prototype.hasOwnProperty;
 
+    /**
+     * Creates an Map<T> for string-based lookups.
+     *
+     * @param slow Creates a slow Map instead of a fast Map. See toSlowMap for more information.
+     */
+    export function createMap<T>(slow?: boolean, template?: Map<T>): Map<T> {
```
Comment (weswigham) [inline src/compiler/core.ts]: This also isn't called anywhere in this PR? Intentional?

File: src/compiler/core.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/core.ts]: Partially. This is something I am currently testing.

**Comment by rbuckton** [comment]
These changes were made in master and merged back into transforms.

**Pull request closed without merging**